File: desktop/src/com/inkostilation/pong/desktop/controls/InputSystem.java
package com.inkostilation.pong.desktop.controls;

import com.badlogic.gdx.Input;
import com.badlogic.gdx.InputAdapter;
import com.inkostilation.pong.engine.Direction;

import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.Set;

public class InputSystem extends InputAdapter {

    private Direction direction = Direction.IDLE;
    private Set<Direction> set = new LinkedHashSet<>();

    private boolean changed = false;

    @Override
    public boolean keyDown(int keycode) {
        switch (keycode) {
            case Input.Keys.UP:
                set.add(Direction.UP);
                changed = true;
                break;
            case Input.Keys.DOWN:
                set.add(Direction.DOWN);
                changed = true;
                break;
        }
        return super.keyDown(keycode);
    }

    @Override
    public boolean keyUp(int keycode) {
        switch (keycode) {
            case Input.Keys.UP:
                set.remove(Direction.UP);
                changed = true;
                break;
            case Input.Keys.DOWN:
                set.remove(Direction.DOWN);
                changed = true;
                break;
        }
        return super.keyUp(keycode);
    }

    public Direction getDirection() {
        if (set.size() > 0) {
            Direction[] dir = new Direction[set.size()];
            set.toArray(dir);
            return dir[set.size() - 1];
        }
        changed = false;
        return Direction.IDLE;
    }

    public boolean isChanged() {
        return changed;
    }
}

File: desktop/src/com/inkostilation/pong/desktop/controls/InputSystem.kt
package com.inkostilation.pong.desktop.controls

import com.badlogic.gdx.Input
import com.badlogic.gdx.InputAdapter
import com.inkostilation.pong.engine.Direction
import java.util.*

class InputSystem : InputAdapter() {
    val direction = Direction.IDLE
        get() {
            if (set.size > 0) {
                val dir = arrayOfNulls<Direction>(set.size)
                set.toArray(dir)
                return dir[set.size - 1]!!
            }
            isChanged = false
            return Direction.IDLE
        }
    private val set: MutableSet<Direction> = LinkedHashSet()
    var isChanged = false
        private set

    override fun keyDown(keycode: Int): Boolean {
        when (keycode) {
            Input.Keys.UP -> {
                set.add(Direction.UP)
                isChanged = true
            }
            Input.Keys.DOWN -> {
                set.add(Direction.DOWN)
                isChanged = true
            }
        }
        return super.keyDown(keycode)
    }

    override fun keyUp(keycode: Int): Boolean {
        when (keycode) {
            Input.Keys.UP -> {
                set.remove(Direction.UP)
                isChanged = true
            }
            Input.Keys.DOWN -> {
                set.remove(Direction.DOWN)
                isChanged = true
            }
        }
        return super.keyUp(keycode)
    }

}